include/picongpu/plugins/saxs/Saxs.kernel

```diff
@@ -0,0 +1,291 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <cstdlib>
+#include <fstream>
+#include <iostream>
+#include <string>
+
+#include "picongpu/simulation_defines.hpp"
+
+#include <pmacc/dimensions/DataSpaceOperations.hpp>
+#include <pmacc/mappings/kernel/AreaMapping.hpp>
+
+#include <pmacc/mappings/threads/ForEachIdx.hpp>
+#include <pmacc/mappings/threads/IdxConfig.hpp>
+#include <pmacc/memory/Array.hpp>
+#include <pmacc/memory/shared/Allocate.hpp>
+#include <pmacc/mpi/MPIReduce.hpp>
+#include <pmacc/mpi/reduceMethods/Reduce.hpp>
+#include <pmacc/nvidia/atomic.hpp>
+#include <pmacc/nvidia/functors/Add.hpp>
+
+#include "sys/stat.h"
+
+namespace picongpu
+{
+
+/** calculate the scattering of a species
+ *
+ * @tparam T_numWorkers number of workers
+ */
+template <uint32_t T_numWorkers> struct KernelSaxs
+{
+    /**
+     * The SAXS kernel calculates for all particles on the device the
+     * scattering intensity for input calculation ranges.
+     * The parallelization is as follows:
+     *  - The number of threads per block is equal to the number of cells per
+     *    super cells which is also equal to the number of particles per frame
+     *
+     * The procedure starts with calculating unique ids for the threads and
+     * initializing the shared memory.
+     * Then a loop over all super cells starts.
+     * Every thread loads a particle from that super cell and calculates its
+     * scattering structure factor.
+     * For every Particle
+     * exists therefore a unique space within the shared memory.
+     * After that, a thread calculates for a specific scattering vector of all
+     * particles.
+     */
+    template <
+        typename ParBox,
+        typename DBox,
+        typename DBox_np,
+        typename DBox_nmp,
+        typename Mapping,
+        typename T_Acc>
+    DINLINE
+    void
+    operator()(
+        T_Acc const &acc,
+        ParBox pb,
+        DBox sumfcoskr,
+        DBox sumfsinkr,
+        DBox_np np,
+        DBox_nmp nmp,
+        DataSpace<simDim> globalOffset,
+        uint32_t currentStep,
+        Mapping mapper,
+        DataSpace<simDim> simBoxSize,
+        float3_64 q_min,
+        float3_64 q_max,
+        float3_64 q_step,
+        unsigned int n_qx,
+        unsigned int n_qy,
+        unsigned int n_qz,
+        unsigned int n_q
+    ) const
+    {
+        using namespace mappings::threads;
+        namespace po = boost::program_options;
+
+        constexpr uint32_t frameSize =
+            pmacc::math::CT::volume<SuperCellSize>::type::value;
+        constexpr uint32_t numWorker = T_numWorkers;
+
+        using FrameType = typename ParBox::FrameType;
+        using FramePtr = typename ParBox::FramePtr;
+
+        uint32_t const workerIdx = threadIdx.x;
```

Comment:
Suggested change of parallelism(2):
Replace this line with `uint32_t const workerIdx = blockIdx.x * blockDim.x + threadIdx.x;` to get a thread index among all blocks (instead of among the current block).